feat(cast): erc20 safe preflight

Syzygy106 opened the pull request:
<!--
Thank you for your Pull Request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests.

Contributors guide: https://github.com/foundry-rs/foundry/blob/HEAD/CONTRIBUTING.md

The contributors guide includes instructions for running rustfmt and building the
documentation.
-->

<!-- ** Please select "Allow edits from maintainers" in the PR Options ** -->

## Motivation

<!--
Explain the context and why you're making that change. What is the problem
you're trying to solve? In some cases there is not a problem and this can be
thought of as being the motivation for your change.
-->

Closes #12402 

Enhance UX and prevent accidental transfers with incorrect amounts. Users may misunderstand the token's decimal places and accidentally send far more (or less) tokens than intended.

Also, approve() requires identical feature, so I decided to add the same.

At the other hand, transferFrom() - doesn't need that, because this method is rarely called by users implicitly, more often him calls other smartcontracts (like swapRouter).

## Solution

<!--
Summarize the solution and provide any necessary context needed to understand
the code change.
-->

Added interactive confirmation prompts to cast erc20 transfer and cast erc20 approve commands that:
1) Fetch token metadata (symbol and decimals) from the ERC20 contract

2) Display human-readable amounts in the confirmation prompt:
Example: Confirm transfer of 100 USDC to address 0x666...666 instead of raw 100000000

3) Provide --yes flag to skip confirmation for non-interactive usage (scripts, CI/CD)
4) Handle edge cases gracefully:
-Falls back to raw amount if decimals/symbol can't be fetched
-Shows warning if token metadata is unavailable

### Demonstration

Let's create a test token first:
```
forge create crates/cast/tests/fixtures/TestToken.sol:TestToken   --private-key 0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80   --rpc-url http://localhost:8545   --broadcast

```

Then we can test behavior:
```
./target/debug/cast erc20 transfer   0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512   0x70997970C51812dc3A010C7d01b50e0d17dc79C8   100000000000000000000   --rpc-url http://localhost:8545   --private-key 0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80

---
Confirm transfer of 100 TEST to address 0x70997970C51812dc3A010C7d01b50e0d17dc79C8 [y/n] 
```

In case y (yes):
```
Confirm transfer of 100 TEST to address 0x70997970C51812dc3A010C7d01b50e0d17dc79C8 yes

---
0x6a6b22e5ff88c63d47527d7d8ff35c3d412388ba7392b1bbc6985f30ccf0e87c

```

In case n (no):
```
Confirm transfer of 100 TEST to address 0x70997970C51812dc3A010C7d01b50e0d17dc79C8 no

---
Error: Transfer cancelled by user
```

With skip promt that would be:
```
./target/debug/cast erc20 transfer \
  0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512 \
  0x70997970C51812dc3A010C7d01b50e0d17dc79C8 \
  50000000000000000000 \
  --yes \
  --rpc-url http://localhost:8545 \
  --private-key 0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80


---

0x3de20e9150169dfe8e7a0c40ecbba21d99587337c3de52d864cd871ed431c0e7

```

In case where contract address aren't correct/contract doesn't support ERC20 interface for decimals/symbol, we'll get a warning:
```
./target/debug/cast erc20 transfer \
  0x1234567890123456789012345678901234567890 \
  0x70997970C51812dc3A010C7d01b50e0d17dc79C8 \
  100 \
  --rpc-url http://localhost:8545 \
  --private-key 0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80


Warning: Warning: Could not fetch token metadata (decimals/symbol). The address may not be a valid ERC20 token contract.
Confirm transfer of 100 TOKEN (raw amount) to address 0x70997970C51812dc3A010C7d01b50e0d17dc79C8 [y/n]
```

The same behavior has been added to approve():
```
./target/debug/cast erc20 approve \
  0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512 \
  0x70997970C51812dc3A010C7d01b50e0d17dc79C8 \
  75000000000000000000 \
  --rpc-url http://localhost:8545 \
  --private-key 0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80

Confirm approval for 0x70997970C51812dc3A010C7d01b50e0d17dc79C8 to spend 75 TEST from your account [y/n]

```


## PR Checklist

- [x] Added Tests
- [x] Added Documentation
- [ ] Breaking changes

Syzygy106 committed:
feat(cast): confirmation for transfer/approve erc20

Syzygy106 committed:
chore(provider): result of clippy cleanup

grandizzy reviewed (commented):
thank you, left a comment re new arg, don't think we need it?

Syzygy106 commented on crates/cast/src/cmd/erc20.rs:
```diff
@@ -65,9 +69,16 @@ pub enum Erc20Subcommand {
+        /// By default, the command will prompt for confirmation before sending the transaction.
+        /// Use this flag to skip the prompt for scripts and non-interactive usage.
+        #[arg(long, short)]
+        yes: bool,
```
> how I was thinking it could be circumvented by `yes | ` instead a new arg

Hmmm, also is possible. But I feel `cast send --yes` more convenient than `yes | cast send`. It looks like other forge/cast style and more suitable for CI/CD as I see.

0xrusowsky replied:
i agree with @grandizzy, the lower amou t of args we maintain, the better.

i feel like the pipe works well enough if we simply document it both in the code and the docs

Syzygy106 committed:
fix(cast): replace --yes flag with yes |

Syzygy106 committed:
chore(cast): delete unused dialoguer from cast

Syzygy106 committed:
chore(cast): linter changes

Syzygy106 committed:
chore(cast): fix docstrings for autodocs

grandizzy reviewed (changes requested):
thank you, left comment, pls check

Syzygy106 commented on crates/cast/src/cmd/erc20.rs:
```diff
@@ -306,21 +347,139 @@ impl Erc20Subcommand {
+                let spender_addr = spender.resolve(&provider).await?;
                 let amount = U256::from_str(&amount)?;
 
+                // Try to fetch token metadata for better UX
```
yep,  you're right

Syzygy106 committed:
chore(cast): compress duplicated code

onbjerg reviewed (changes requested):
i know you can pipe in `yes`, but let's also add a `--yes/-y` as this is pretty common for clis

Syzygy106 commented on crates/cast/src/cmd/erc20.rs:
```diff
@@ -15,6 +18,84 @@ use foundry_wallets::WalletOpts;
+                    }
+                } else {
+                    formatted
+                };
```
> not really sure what the trim routine is for?

The trim routine removes unnecessary trailing zeros and decimal points from the formatted amount to make it more human-readable.

Without trimming:
100.000000000000000000 TEST (18 decimals)
1.500000000000000000 TEST

With trimming:
100 TEST
1.5 TEST

I feel that more convenient UX

Syzygy106 commented:
> i know you can pipe in `yes`, but let's also add a `--yes/-y` as this is pretty common for clis

Hi there, @onbjerg , I actually agree with the `--yes` flag approach - that was my initial implementation (you can see it in the PR history above - https://github.com/foundry-rs/foundry/pull/12499#discussion_r2527576406).

However, other maintainers had different perspectives:
1) @grandizzy (the original issue author) suggested removing the `--yes` flag, saying "don't think we need it?" https://github.com/foundry-rs/foundry/pull/12499#discussion_r2526285635
2) @0xrusowsky also seemed to prefer the pipe approach https://github.com/foundry-rs/foundry/pull/12499#discussion_r2536408180

Their reasoning was avoiding adding new arguments to keep the CLI simpler and easier to maintain and following Unix philosophy with `yes |` pipe being a standard pattern.

I'm happy to implement either approach, but before making changes again, I think it would be helpful if the maintainers could align on the preferred solution:

Option 1: `--yes/-y` flag (more explicit, common in modern CLIs)  (I prefer this)
Option 2: `yes |` pipe only (Unix philosophy, no new flags)

@grandizzy @0xrusowsky @onbjerg - could you please discuss and let me know which direction you'd like me to take? I want to make sure we're all aligned before implementing the final version.

Other maintainers are also invited to the conversation @mattsse , @zerosnacks , @DaniPopes

grandizzy commented:
@Syzygy106 sorry for back and forth, we'll discuss and make a final call re how to add this. thank you!

Syzygy106 commented:
> @Syzygy106 sorry, my bad, should have been get full consensus on it. Let's stick with your initial impl, that is adding the --yes/-y arg. Thank you

Yep, no problem!

Syzygy106 committed:
restore(cast): return to erc20 --yes flag (223bf2da)

Syzygy106 committed:
chore(cast): remove duplicate via helper

DaniPopes committed:
Merge branch 'master' into feat/cast-erc20-safe_preflight

gakonst committed:
Merge origin/master into pr-12499 - resolve conflicts

Merged confirmation prompt feature (--yes flag) with master's updates:
- Combined --yes flag for transfer/approve with curl support and tx_opts
- Added --yes flag to all test cases that use transfer/approve
- Kept both PR's confirmation tests and master's gas/curl tests

Amp-Thread-ID: https://ampcode.com/threads/T-019bdbc8-2e58-7396-9320-bbd9c1933a27
Co-authored-by: Amp <amp@ampcode.com>

gakonst committed:
fix: update tests to not check output format

The tests were checking that output starts with 0x (tx hash), but
cast_send outputs the full receipt in sync mode. Updated tests to
just verify the transaction succeeded via balance/allowance checks.

emmajam commented:
hey we are doing some spring cleaning and closing old pull requests, please reopen this if you think its still needed and valuable, thank you for your contributions! ❤️

Syzygy106 commented:
@grandizzy , @0xrusowsky there almost merged too. Could we reopen and merge?
